pkg/controllers/provisioner/controller.go

```diff
@@ -0,0 +1,72 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioner
+
+import (
+	"context"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	"go.uber.org/multierr"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Controller struct {
+	kubeClient client.Client
+}
+
+func NewController(kubeClient client.Client) corecontroller.Controller {
+	return corecontroller.Typed[*v1alpha5.Provisioner](kubeClient, &Controller{
+		kubeClient: kubeClient,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "provisioner_hash"
+}
+
+// Reconcile the resource
+func (c *Controller) Reconcile(ctx context.Context, p *v1alpha5.Provisioner) (reconcile.Result, error) {
+	stored := p.DeepCopy()
+
+	provisionerHash, err := p.StaticDriftHash()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if p.ObjectMeta.Annotations == nil {
+		p.ObjectMeta.Annotations = map[string]string{v1alpha5.ProvisionerHashAnnotationKey: provisionerHash}
+	} else {
+		p.ObjectMeta.Annotations[v1alpha5.ProvisionerHashAnnotationKey] = provisionerHash
+	}
+
+	if patchErr := c.kubeClient.Patch(ctx, p, client.MergeFrom(stored)); patchErr != nil {
```

Comment:
Can we check if stored is different than p after the reconcile? Will save on patches if we don't need to.